File: sdk/dart-sdk/test/unit/remoteconfig_test.dart
import 'package:test/test.dart';
import 'package:ownfirebase_sdk/types.dart';
import 'package:ownfirebase_sdk/remoteconfig.dart';

void main() {
  group('RemoteConfigSDK Unit Tests (Mocked)', () {
    late RemoteConfigSDK remoteConfig;
    late OwnFirebaseConfig config;

    setUp(() {
      config = OwnFirebaseConfig(
        baseUrl: 'http://localhost:8000',
        projectId: 'test-project',
      );
      remoteConfig = RemoteConfigSDK(config: config);
    });

    test('RemoteConfigSDK initializes with config', () {
      expect(remoteConfig.baseUrl, equals('http://localhost:8000'));
      expect(remoteConfig.projectId, equals('test-project'));
    });

    test('RemoteConfigParameter.fromJson parses correctly', () {
      final json = {
        'id': 'param-123',
        'key': 'feature_flag_new_ui',
        'default_value': 'false',
        'description': 'Enable new UI version',
        'value_type': 'boolean',
      };
      final param = RemoteConfigParameter.fromJson(json);
      expect(param.id, equals('param-123'));
      expect(param.key, equals('feature_flag_new_ui'));
      expect(param.defaultValue, equals('false'));
      expect(param.valueType, equals('boolean'));
    });

    test('RemoteConfigParameter with string type', () {
      final json = {
        'id': 'param-1',
        'key': 'app_version',
        'default_value': '1.0.0',
        'description': 'Current app version',
        'value_type': 'string',
      };
      final param = RemoteConfigParameter.fromJson(json);
      expect(param.valueType, equals('string'));
      expect(param.defaultValue, equals('1.0.0'));
    });

    test('RemoteConfigParameter with number type', () {
      final json = {
        'id': 'param-2',
        'key': 'max_retry_count',
        'default_value': '3',
        'description': 'Maximum retry count',
        'value_type': 'number',
      };
      final param = RemoteConfigParameter.fromJson(json);
      expect(param.valueType, equals('number'));
      expect(param.defaultValue, equals('3'));
    });

    test('RemoteConfigParameter with json type', () {
      final json = {
        'id': 'param-3',
        'key': 'feature_config',
        'default_value': '{"enabled":true,"level":2}',
        'description': 'Feature configuration',
        'value_type': 'json',
      };
      final param = RemoteConfigParameter.fromJson(json);
      expect(param.valueType, equals('json'));
      expect(param.defaultValue, contains('{'));
    });

    test('Parameter listing pagination', () {
      final response = {
        'count': 25,
        'next': 'http://localhost:8000/api/projects/test/remote-config/parameters/?page=2',
        'previous': null,
        'results': [
          {
            'id': 'param-1',
            'key': 'feature_1',
            'default_value': 'true',
            'description': 'Feature 1',
            'value_type': 'boolean',
          },
          {
            'id': 'param-2',
            'key': 'feature_2',
            'default_value': 'false',
            'description': 'Feature 2',
            'value_type': 'boolean',
          },
        ],
      };
      expect(response['count'], equals(25));
      expect(response['results'], hasLength(2));
    });

    test('Fetch parameters structure', () {
      final fetchResult = {
        'feature_flag_new_ui': 'true',
        'max_retry_count': '5',
        'app_version': '2.0.0',
      };
      expect(fetchResult['feature_flag_new_ui'], equals('true'));
      expect(fetchResult.keys, hasLength(3));
    });

    test('Fetch with context structure', () {
      final context = {
        'user_id': 'user-123',
        'app_version': '1.5.0',
        'platform': 'ios',
      };
      expect(context['user_id'], equals('user-123'));
      expect(context['platform'], equals('ios'));
    });

    test('Parameter value types all supported', () {
      final types = ['string', 'boolean', 'number', 'json'];
      expect(types, hasLength(4));
      expect(types.contains('string'), isTrue);
      expect(types.contains('json'), isTrue);
    });

    test('Multiple parameters creation', () {
      final params = [
        {'key': 'param1', 'default_value': 'value1'},
        {'key': 'param2', 'default_value': 'value2'},
        {'key': 'param3', 'default_value': 'value3'},
      ];
      expect(params, hasLength(3));
      expect(params.map((p) => p['key']).toList(), contains('param1'));
    });

    test('Parameter update operation', () {
      final updatePayload = {
        'default_value': 'new_value',
        'description': 'Updated description',
      };
      expect(updatePayload['default_value'], equals('new_value'));
      expect(updatePayload.keys, hasLength(2));
    });
  });
}

File: sdk/__tests__/remoteconfig.test.ts
import { RemoteConfigSDK } from '../src/remoteconfig';
import type { RemoteConfigParameter, PaginatedResponse } from '../src/types';
import type { ConfigCondition } from '../src/remoteconfig';

global.fetch = jest.fn();

describe('RemoteConfigSDK', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const config = {
    baseUrl: 'http://localhost:8000',
    projectId: 'test-project',
  };

  describe('parameters', () => {
    it('should list parameters', async () => {
      const mockResponse: PaginatedResponse<RemoteConfigParameter> = {
        count: 2,
        next: null,
        previous: null,
        results: [
          {
            id: 'param1',
            key: 'feature_flag_new_ui',
            default_value: 'false',
            description: 'Enable new UI design',
            value_type: 'boolean',
          },
          {
            id: 'param2',
            key: 'api_timeout_ms',
            default_value: '5000',
            description: 'API request timeout in milliseconds',
            value_type: 'number',
          },
        ],
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockResponse,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.listParameters();

      expect(result.count).toBe(2);
      expect(result.results).toHaveLength(2);
    });

    it('should get specific parameter', async () => {
      const mockParameter: RemoteConfigParameter = {
        id: 'param1',
        key: 'feature_flag_new_ui',
        default_value: 'false',
        description: 'Enable new UI design',
        value_type: 'boolean',
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockParameter,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.getParameter('param1');

      expect(result).toEqual(mockParameter);
      expect(result.key).toBe('feature_flag_new_ui');
    });

    it('should create parameter', async () => {
      const mockParameter: RemoteConfigParameter = {
        id: 'param3',
        key: 'app_version_min',
        default_value: '1.0.0',
        description: 'Minimum app version',
        value_type: 'string',
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 201,
        json: async () => mockParameter,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.createParameter({
        key: 'app_version_min',
        default_value: '1.0.0',
        description: 'Minimum app version',
        value_type: 'string',
      });

      expect(result).toEqual(mockParameter);
      expect(global.fetch).toHaveBeenCalledWith(
        'http://localhost:8000/api/projects/test-project/config/parameters/',
        expect.objectContaining({
          method: 'POST',
        })
      );
    });

    it('should update parameter', async () => {
      const mockParameter: RemoteConfigParameter = {
        id: 'param1',
        key: 'feature_flag_new_ui',
        default_value: 'true',
        description: 'Enable new UI design (updated)',
        value_type: 'boolean',
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockParameter,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.updateParameter('param1', {
        default_value: 'true',
        description: 'Enable new UI design (updated)',
      });

      expect(result.default_value).toBe('true');
      expect(result.description).toContain('updated');
    });

    it('should delete parameter', async () => {
      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 204,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.deleteParameter('param1');

      expect(result).toBeUndefined();
    });

    it('should handle different value types', async () => {
      const stringParam: RemoteConfigParameter = {
        id: 'p1',
        key: 'env',
        default_value: 'production',
        description: 'Environment',
        value_type: 'string',
      };

      const boolParam: RemoteConfigParameter = {
        id: 'p2',
        key: 'debug_mode',
        default_value: 'false',
        description: 'Debug mode',
        value_type: 'boolean',
      };

      const numberParam: RemoteConfigParameter = {
        id: 'p3',
        key: 'max_retries',
        default_value: '3',
        description: 'Max retries',
        value_type: 'number',
      };

      const jsonParam: RemoteConfigParameter = {
        id: 'p4',
        key: 'feature_config',
        default_value: '{"enabled": true}',
        description: 'Feature config',
        value_type: 'json',
      };

      expect(stringParam.value_type).toBe('string');
      expect(boolParam.value_type).toBe('boolean');
      expect(numberParam.value_type).toBe('number');
      expect(jsonParam.value_type).toBe('json');
    });
  });

  describe('conditions', () => {
    it('should list conditions for parameter', async () => {
      const mockConditions: ConfigCondition[] = [
        {
          id: 'cond1',
          name: 'ios_users',
          expression: 'user.platform == "ios"',
          value: 'true',
        },
        {
          id: 'cond2',
          name: 'beta_testers',
          expression: 'user.email like "%@beta.com"',
          value: 'true',
        },
      ];

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockConditions,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.listConditions('param1');

      expect(result).toEqual(mockConditions);
      expect(result).toHaveLength(2);
    });

    it('should create condition for parameter', async () => {
      const mockCondition: ConfigCondition = {
        id: 'cond3',
        name: 'android_users',
        expression: 'user.platform == "android"',
        value: 'false',
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 201,
        json: async () => mockCondition,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.createCondition('param1', {
        name: 'android_users',
        expression: 'user.platform == "android"',
        value: 'false',
      });

      expect(result).toEqual(mockCondition);
      expect(global.fetch).toHaveBeenCalledWith(
        'http://localhost:8000/api/projects/test-project/config/parameters/param1/conditions/',
        expect.objectContaining({
          method: 'POST',
        })
      );
    });

    it('should update condition', async () => {
      const mockCondition: ConfigCondition = {
        id: 'cond1',
        name: 'ios_users',
        expression: 'user.platform == "ios" && user.app_version >= "2.0"',
        value: 'true',
      };

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 200,
        json: async () => mockCondition,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.updateCondition('param1', 'cond1', {
        expression: 'user.platform == "ios" && user.app_version >= "2.0"',
      });

      expect(result.expression).toContain('app_version');
    });

    it('should delete condition', async () => {
      (global.fetch as jest.Mock).mockResolvedValueOnce({
        ok: true,
        status: 204,
      });

      const remoteConfig = new RemoteConfigSDK(config);
      remoteConfig.setAccessToken('access-token');
      remoteConfig.setProjectId('test-project');

      const result = await remoteConfig.deleteCondition('param1', 'cond1');

      expect(result).toBeUndefined();
      expect(global.fetch).toHaveBeenCalledWith(
        'http://localhost:8000/api/projects/test-project/config/parameters/param1/conditions/cond1/',
        expect.objectContaining({
          method: 'DELETE',
        })
      );
    });

    it('should handle complex condition expressions', async () => {
      const complexConditions: ConfigCondition[] = [
        {
          id: 'cond-complex1',
          name: 'premium_users_us',
          expression: 'user.subscription_tier == "premium" && user.country == "US" && user.created_at > "2024-01-01"',
          value: '{"feature": "enabled", "tier": "premium"}',
        },
        {
          id: 'cond-complex2',
          name: 'beta_rollout_10_percent',
          expression: 'user.id % 100 < 10',
          value: 'true',
        },
      ];

      expect(complexConditions[0].expression).toContain('&&');
      expect(complexConditions[1].expression).toContain('%');
    });

    it('should handle condition overrides', async () => {
      const baseParam: RemoteConfigParameter = {
        id: 'param-base',
        key: 'feature_enabled',
        default_value: 'false',
        description: 'Base feature flag',
        value_type: 'boolean',
      };

      const conditionOverrides: ConfigCondition[] = [
        {
          id: 'override1',
          name: 'internal_team',
          expression: 'user.email like "%@company.com"',
          value: 'true',
        },
        {
          id: 'override2',
          name: 'beta_program',
          expression: 'user.in_beta_program == true',
          value: 'true',
        },
      ];

      expect(baseParam.default_value).toBe('false');
      expect(conditionOverrides[0].value).toBe('true');
      expect(conditionOverrides[1].value).toBe('true');
    });
  });

  describe('config rollout scenarios', () => {
    it('should support gradual rollout with percentage', async () => {
      const rolloutCondition: ConfigCondition = {
        id: 'rollout-10',
        name: 'gradual_rollout_10_percent',
        expression: 'user.id % 100 < 10',
        value: 'true',
      };

      expect(rolloutCondition.expression).toContain('% 100 < 10');
    });

    it('should support geo-targeted config', async () => {
      const geoCondition: ConfigCondition = {
        id: 'geo-us',
        name: 'us_only_feature',
        expression: 'user.country in ["US", "CA"]',
        value: '{"enabled": true, "variant": "us_version"}',
      };

      expect(geoCondition.expression).toContain('country');
      expect(geoCondition.value).toContain('variant');
    });

    it('should support version-based config', async () => {
      const versionCondition: ConfigCondition = {
        id: 'version-check',
        name: 'minimum_version_requirement',
        expression: 'user.app_version >= "2.0.0"',
        value: '{"min_version": "2.0.0"}',
      };

      expect(versionCondition.expression).toContain('app_version');
    });
  });
});
